linode/databasebackups/framework_datasource.go

```diff
@@ -0,0 +1,112 @@
+package databasebackups
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/linode/linodego"
+	"github.com/linode/terraform-provider-linode/linode/helper"
+)
+
+func NewDataSource() datasource.DataSource {
+	return &DataSource{}
+}
+
+type DataSource struct {
+	client *linodego.Client
+}
+
+func (r *DataSource) Configure(
+	ctx context.Context,
+	req datasource.ConfigureRequest,
+	resp *datasource.ConfigureResponse,
+) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	meta := helper.GetDataSourceMeta(req, resp)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	r.client = meta.Client
+}
+
+func (r *DataSource) Metadata(
+	ctx context.Context,
+	req datasource.MetadataRequest,
+	resp *datasource.MetadataResponse,
+) {
+	resp.TypeName = "linode_database_backups"
+}
+
+func (r *DataSource) Schema(
+	ctx context.Context,
+	req datasource.SchemaRequest,
+	resp *datasource.SchemaResponse,
+) {
+	resp.Schema = frameworkDataSourceSchema
+}
+
+func (r *DataSource) Read(
+	ctx context.Context,
+	req datasource.ReadRequest,
+	resp *datasource.ReadResponse,
+) {
+	var data DatabaseBackupFilterModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	data.ID = data.DatabaseID
+
+	listMySQLBackups := func(ctx context.Context, client *linodego.Client, filter string) ([]any, error) {
+		backups, err := client.ListMySQLDatabaseBackups(ctx, int(data.DatabaseID.ValueInt64()), &linodego.ListOptions{
+			Filter: filter,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return helper.TypedSliceToAny(backups), nil
+	}
+
+	listPostgresSQLBackups := func(ctx context.Context, client *linodego.Client, filter string) ([]any, error) {
+		backups, err := client.ListPostgresDatabaseBackups(ctx, int(data.DatabaseID.ValueInt64()), &linodego.ListOptions{
+			Filter: filter,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return helper.TypedSliceToAny(backups), nil
+	}
```

Comment:
This is an optional suggestion.
I feel moving the function definition into the if else block and only define listMySQLBackups when the db type is mysql might improve the memory usage slightly, what do you think?

Reply:
Good idea. I implemented this change!